fw/http_frame.c

```diff
@@ -598,20 +613,11 @@ __tfw_h2_stream_unlink(TfwH2Ctx *ctx, TfwStream *stream)
 		--ctx->hclosed_streams.num;
 	}
 
-	if (hmreq) {
+	if (hmreq)
 		hmreq->stream = NULL;
-		/*
-		 * If the request is linked with a stream, but not complete yet,
-		 * it must be deleted right here to avoid leakage, because in
-		 * this case it is not used anywhere yet. When request is
-		 * assembled and complete, it will be removed (due to some
-		 * processing error) in @tfw_http_req_process(), or in other
-		 * cases controlled by server connection side (after adding to
-		 * @fwd_queue): successful response sending, eviction etc.
-		 */
-		if (!test_bit(TFW_HTTP_B_FULLY_PARSED, hmreq->flags))
-			tfw_http_conn_msg_free(hmreq);
-	}
+
+	/* Note that we may have not completed requests when websockets upgrade,
+	 * so we should not free msg on not TFW_HTTP_B_FULLY_PARSED messages. */
```

Comment:
That was special code path for no completed yet requests. Requests always freed in `tfw_http_resp_pair_free()`, for example, on response forward in `fw/http.c:956` or on client connection drop `fw/http.c:2803`.

Reply:
OK, then what about not complete and not websocket requests? Where they are freed? Or did we have a double free bug?

Also I'm confused with `fw/http.c:956` case since now in `tfw_h2_resp_fwd()` we free response only for websocket connections, request isn't freed.